csharp/src/Apache.Arrow/C/CArrowArrayExporter.cs

```diff
@@ -184,13 +189,12 @@ private unsafe static void ConvertRecordBatch(ExportedAllocationOwner sharedOwne
             cArray->dictionary = null;
         }
 
+#if NET5_0_OR_GREATER
+        [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvStdcall) })]
```

Comment:
See https://github.com/apache/arrow/pull/34133#discussion_r1122532439. Ideally we would have used the default calling convention, but that would not be suppported on anything earlier than .NET 5. And in 64-bit platforms the calling convention doesn't matter either way.

Reply:
Well, the entire point (mostly) of implementing the C Data Interface is to be compatible with non-.Net producers/consumers. Those are extremely likely to use the platform default. So we should get it right at least when possible, i.e. on .Net >= 5.0.

As for https://stackoverflow.com/questions/34832679/is-the-callingconvention-ignored-in-64-bit-net-applications , does it apply here? It's talking about `DllImport`, which might be different from this?

Reply:
In any case, keeping the default convention seems more theoretically sound (and forward-looking, perhaps).

Reply:
cc @westonpace @lidavidm for opinions.

Reply:
If I understand correctly, are you saying to change this: https://github.com/apache/arrow/blob/e628ca50078c749ea7a10b53defe12cbec7d581f/csharp/src/Apache.Arrow/C/CArrowArray.cs#L41

into this?

```csharp
#if NET5_0_OR_GREATER
public delegate* unmanaged<CArrowArray*, void> release;
#else
public delegate* unmanaged[Stdcall]<CArrowArray*, void> release;
#endif
```

That would cause an incompatible API surface between the assembly compiled for .NET 6 and that compiled for the earlier frameworks. We have two options:

* Lie and keep the `stdcall` calling convention on the function pointers.
* Use the default unmanaged calling convention but support the C interface only on .NET 6+ (we don't target 5 as it is unsupported).

Reply:
If you take a look for example at the  Arrow C++ implementation, its release functions are entirely private. For example `ReleaseExportedSchema` below is in the anonymous namespace, which doesn't expose the function publicly:
https://github.com/apache/arrow/blob/e628ca50078c749ea7a10b53defe12cbec7d581f/cpp/src/arrow/c/bridge.cc#L107

Reply:
So I should make the members of these structs private? That also works.